### DISCUSSION TITLE
抗 TCP connection reset 的代理协议 TCP over TCP

### DISCUSSION THREAD

**Discussion by ghost** [v2fly/v2ray-core#2687] [Ideas]
我在一个UDP被完全block的网络中，有时候GFW会莫名其妙导致 TCP connection reset 。

当代理的TCP链接断开时，
用户浏览器 ---> 代理客户端 --x--> 代理服务器  ---> 目标网站

目前绝大部分代理协议的行为是，直接把TCP reset传到应用层app。这样在一些不支持断点续传的软件中，容易引发痛苦
用户浏览器 -x-> 代理客户端 --x--> 代理服务器  -x-> 目标网站

建议在代理协议（已加密的数据）中，插入显式的“用户浏览器TCP断开“和”目标网站TCP断开”信号，在代理客户端收到”目标网站TCP断开”信号前，即使代理TCP链接断开，都会保持用户侧链接的开启，并重新启动新的代理TCP链接来延续上一个断掉的代理TCP链接。

用户浏览器 --缓存--> 代理客户端 --x--> 代理服务器  --缓存--> 目标网站
用户浏览器 ---> 代理客户端 --新链接--> 代理服务器  ---> 目标网站

同理，服务器端收到“用户浏览器TCP断开“前，会在一个合理的超时时间内，保持通往目标网站的TCP链接的开启，直到下行缓存区满。

另外，启动新连接时应当发送与上个被延续的链接相同的识别码（如相同的随机数或UUID），代理服务器根据该识别码确定恢复哪个链接。

**Comment by ghost** [comment]
@RPRX @xiaokangwang @nekohasekai

**Reply by RPRX** [reply]
~~好吧既然 @ 我了~~，TCP 上的连接迁移，增强版 XUDP 已经有了（意外中断，UDP 不用缓存），Switch 正在路上（主动迁移，非意外中断）

我觉得做意外中断的 TCP 迁移有点麻烦，要缓存要包序号要超时重拨策略等，不如直接用 QUIC 类代理协议，~~或者 QUIC over TCP~~

**Reply by ghost** [reply]
但QUIC 主要问题是UDP协议本身被丢包严重，而且很多网络环境根本就是block udp的（比如我）。

用UDP代理当内层+UDP over TCP代理当外层，做成套娃，是能达成类似的效果（甚至还自带mux功效），但套娃有很多header上的带宽开销、而且把TCP转UDP（乱序）再套进TCP，感觉加了好几重不必要的buffer。

如果能保证内层TCP不断的情况下，实现外层TCP链接的重连和迁移，也就是TCP over TCP，应该会带来前所未有的稳定体验

**Reply by RPRX** [reply]
加个对包序号的 ACK 机制对缓存的压力就会小很多，还可以配置多久没收到 ACK 就“快速重连”，总之带 ACK 机制的话应该不难实现

**Reply by ghost** [reply]
现在xray能实现用TCP over UDP代理（某种？）当内层+UDP over TCP代理（XUDP）当外层，做成套娃的形式了么？

**Reply by RPRX** [reply]
以前我遇到过一个喜欢掐连接的运营商，导致我用浏览器下东西虽然支持断点续传，但总是要手动继续，挺烦的，当时也想要这样的东西，~~不过也可以通过自动继续来实现~~。其实应用层基本上都有应对 TCP 断连的策略，你不支持断点续传的是什么场景？

**Reply by ghost** [reply]
s3协议传大文件，还有几个傻逼播放器（）。浏览网页也经常被掐，导致不敢开mux，一掐掐断一堆。

哦，还有ssh，非常烦。

**Reply by RPRX** [reply]
> 现在xray能实现用TCP over UDP代理（某种？）当内层+UDP over TCP代理（XUDP）当外层，做成套娃的形式了么？

TCP 套娃可以，UDP 套娃~~应该也可以~~，不过内部套娃的话状态可能是互绑的，所以建议你走任意门+支持连接迁移的 UDP 协议

**Reply by ghost** [reply]
支持连接迁移的 UDP 协议现在有哪些可以选呢？

**Reply by RPRX** [reply]
> 支持连接迁移的 UDP 协议现在有哪些可以选呢？

QUIC 类

**Reply by RPRX** [reply]
~~不过既然你要放 TCP 里面，建议你直接用 *ray 的上古 QUIC，因为遇到发包激进的认为它堵了反而更麻烦~~

**Reply by ghost** [reply]
上古 QUIC是啥？好多年没用quic了，现在的版本的QUIC学kcp那样激进发包了么？

**Reply by RPRX** [reply]
> 上古 QUIC是啥？好多年没用quic了，现在的版本的QUIC学kcp那样激进发包了么？

不是现在版本的 QUIC，而是一些魔改版 QUIC，当然在表现上肯定与普通 QUIC 有些区别，用的人少，GFW 暂时没管到这个层面

**Reply by ghost** [reply]
TUIC 和 Hysteria有激进发包么？按理说有只有拥塞控制的没事，不重复发包就行？

**Reply by RPRX** [reply]
> TUIC 和 Hysteria有激进发包么？按理说有只有拥塞控制的没事，不重复发包就行？

~~起速快+适合丢包环境，你觉得怎么做到~~

**Reply by ghost** [reply]
那如果套在tcp里用，比如原版quic的ReceiveWindow应该增大还是减小呢？

**Reply by RPRX** [reply]
> 那如果套在tcp里用，比如原版quic的ReceiveWindow应该增大还是减小呢？

这是不断变化的吧，~~没研究过，建议实测~~

**Reply by RPRX** [reply]
> 支持连接迁移的 UDP 协议现在有哪些可以选呢？

对了，如果你用的是增强版 XUDP，它自带 UoT 迁移，就不必要求 UDP 协议支持连接迁移了，任意可靠传输 UDP 协议均可。

**Comment by ghost** [comment]
不用UDP叠床架屋的改法是让代理TCP协议默认自动重连。

一种（不正确的）实现方法：在TCP代理协议的header上加一个随机数作为链接识别码，并加上1个表示动作的标志位（1:打开新连接并设置识别码或继续有相同识别码的某个内层TCP链接，0:断开先前的相同识别码的某个内层TCP链接），然后把客户端代理TCP的默认行为改成“断开后自动重连”。服务器的默认行为改为“在一个合理的超时时间内，保持通往目标网站的TCP链接的开启，直到下行缓存区满”。

这样的坏处是，如果是内层TCP协议断开导致外层TCP断开的情况（真断开），客户端还需要再开一个新的TCP连接告诉代理服务器去断开通往远程的内层TCP链接。如果内层TCP是被远程服务器断开了，代理服务器在收到重连时需要返回“无法继续该识别码对应的TCP链接”。

**Reply by RPRX** [reply]
没有这么简单，要基于不可靠传输做可靠传输，肯定要有缓存和重传机制，~~增强版 XUDP 由于是传输 UDP 所以一个识别码就够了~~

**Reply by ghost** [reply]
但感觉TCP基本上是可靠的？至少不会乱序？

**Reply by RPRX** [reply]
> 但感觉TCP基本上是可靠的？至少不会乱序或丢包？

上一个连接意外断开时很有可能有些数据你发出去了但对方没有收到，所以要有 ACK 机制，对方没 ACK 的都要缓存着，可能要重传

**Reply by qeatzy** [reply]
为了ssh永不掉线，实现过这个，基本上就是复制到cache+发送 -> 接收+send Ack -> recv Ack+forward idx in Cache。Ack后idx没前进的触发重传。Ack没有采用selective Ack，就是简单的seq+packet len（毕竟底层是tcp嘛）。用了mem pool减少gc压力，读写分离各自在不同线程，还没有上线程池（暂无必要）。

**Reply by ghost** [reply]
@qeatzy  求PR，救救孩子

**Reply by qeatzy** [reply]
v2不熟，不敢班门弄斧。

**Comment by ghost** [comment]
另外一种改法是做一个“永生”的、能够支持底层代理链接切换（有分片 ACK 机制）的TCP over TCP。

**Reply by RPRX** [reply]
其实就只能这么改，~~因为你上面那个说法没做到可靠传输~~，有没有 Mux 不重要

**Reply by ghost** [reply]
对，和mux无关，不用要求必须开启mux。我只是想到偷懒的改法可能是改mux。但好像现在的mux都没ack，也不能永生。。

**Comment by AkinoKaede** [comment]
我有一点疑虑，断开连接的指令有没有可能成为特征？如果有，如何设计以解决这个问题？

I have a doubt, is it possible that the disconnect command could become a characteristic? If so, how to design to solve this problem?

**Reply by ghost** [reply]
会。所以第二种（使用TCP over QUIC+ UDP over TCP套娃来实现TCP over TCP的） 方法可能更合适。

但依然有两个小问题：大部分QUIC默认参数设定都不太适合被套在UDP over TCP隧道内。而且似乎除了额外配个127.0.0.1上的端口外，没法把QUIC出站流量塞回另一个代理当入站。

**Reply by AkinoKaede** [reply]
把 IP 包本身作为载荷似乎能更好地解决这个问题，但是这会在实现上造成困难，也会让应用场景受限。

Using IP packets as the payload seems to be a better solution to this problem, but this will cause difficulties in implementation and will also limit the application scenarios.

**Reply by AkinoKaede** [reply]
把 TCP 载荷重组为 IP 包也是可行的，有开发者做过这样的事。

It is also possible to reassemble TCP payloads into IP packets, and some developers have done this.

**Reply by ghost** [reply]
openvpn的tcp模式是这样的，但是routing需要走linux内核的，tun需要root权限，而且没有代理软件配置起来简单、灵活。

**Reply by ghost** [reply]
您说的是类似fakeTCP的方法么？

**Reply by AkinoKaede** [reply]
不是，我指的是在应用层代理软件上实现 WireGuard。

No, I am referring to the implementation of WireGuard on application layer proxy software.

**Reply by AkinoKaede** [reply]
这算是对现有 VPN  协议的修改，还可以通过连接池来进一步提升用户体验，不过 V2Ray 看起来不适合做这样的事。

This is considered a modification to the existing VPN protocols, and can further improve the user experience through connection pooling, but V2Ray does not seem suitable for such a thing.
